google/cloud/alloydb/connector/connector.py

```diff
@@ -149,13 +159,117 @@ async def connect_async(self, instance_uri: str, driver: str, **kwargs: Any) ->
+
+        Args:
+            ip_address (str): IP address of AlloyDB instance to connect to.
+            ctx (ssl.SSLContext): Context used to create a TLS connection
+                with AlloyDB instance ssl certificates.
+            enable_iam_auth (bool): Flag to enable IAM database authentication.
+            driver (str): A string representing the database driver to connect with.
+                Supported drivers are pg8000.
+
+        Returns:
+            sock (ssl.SSLSocket): mTLS/SSL socket connected to AlloyDB Proxy server.
+        """
+        # Create socket and wrap with SSL/TLS context
+        sock = ctx.wrap_socket(
+            socket.create_connection((ip_address, SERVER_PROXY_PORT)),
+            server_hostname=ip_address,
+        )
+        # set auth type for metadata exchange
+        auth_type = connectorspb.MetadataExchangeRequest.DB_NATIVE
+        if enable_iam_auth:
+            auth_type = connectorspb.MetadataExchangeRequest.AUTO_IAM
+
+        # form metadata exchange request
+        req = connectorspb.MetadataExchangeRequest(
+            user_agent=self._client._user_agent + f"+{driver}",  # type: ignore
+            auth_type=auth_type,
+            oauth2_token=self._credentials.token,
+        )
+
+        # set I/O timeout
+        sock.settimeout(30)
+
+        # pack big-endian unsigned integer (4 bytes)
+        packed_len = struct.pack(">I", req.ByteSize())
+
+        # send metadata message length
+        sock.sendall(packed_len)
+        # send metadata request message
+        sock.sendall(req.SerializeToString())
+
+        # form metadata exchange response
+        resp = connectorspb.MetadataExchangeResponse()
+
+        # read metadata message length (4 bytes)
+        message_len_buffer_size = struct.Struct("I").size
+        message_len_buffer = b""
+        while message_len_buffer_size > 0:
+            chunk = sock.recv(message_len_buffer_size)
+            if not chunk:
+                raise RuntimeError(
+                    "Connection closed while getting metadata exchange length!"
+                )
+            message_len_buffer += chunk
+            message_len_buffer_size -= len(chunk)
+
+        (message_len,) = struct.unpack(">I", message_len_buffer)
+
+        # read metadata exchange message
+        buffer = b""
+        while message_len > 0:
+            chunk = sock.recv(message_len)
+            if not chunk:
+                raise RuntimeError(
+                    "Connection closed while performing metadata exchange!"
+                )
+            buffer += chunk
+            message_len -= len(chunk)
+
+        # parse metadata exchange response from buffer
+        resp.ParseFromString(buffer)
+
+        # validate metadata exchange response
+        if resp.response_code != connectorspb.MetadataExchangeResponse.OK:
+            raise ValueError("Metadata Exchange request has failed!")
```

Comment:
Let's add the error message from the response in here.

Reply:
let me know how this looks now, can't mimic the same as Go or Java as the Java and Go proto libraries at helper methods like `GetResponseCode` which I don't have here in PYthon